gcp/src/main/java/org/apache/iceberg/gcp/gcs/GCSOutputStream.java

```diff
@@ -0,0 +1,127 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.gcp.gcs;
+
+import com.google.api.client.util.Lists;
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobWriteOption;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.channels.Channels;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.iceberg.gcp.GCPProperties;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The GCSOutputStream leverages native streaming channels from the GCS API
+ * for streaming uploads. See <a href="https://cloud.google.com/storage/docs/streaming">Streaming Transfers</a>
+ */
+class GCSOutputStream extends PositionOutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(GCSOutputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final Storage storage;
+  private final BlobId blobId;
+  private final GCPProperties gcpProperties;
+
+  private OutputStream stream;
+
+  private long pos = 0;
+  private boolean closed = false;
+
+  GCSOutputStream(Storage storage, BlobId blobId, GCPProperties gcpProperties) throws IOException {
+    this.storage = storage;
+    this.blobId = blobId;
+    this.gcpProperties = gcpProperties;
+
+    createStack = Thread.currentThread().getStackTrace();
+
+    openStream();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void flush() throws IOException {
+    stream.flush();
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    stream.write(b);
+    pos += 1;
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    stream.write(b, off, len);
+    pos += len;
+  }
+
+  private void openStream() {
+    List<BlobWriteOption> writeOptions = Lists.newArrayList();
+
+    gcpProperties.encryptionKey().ifPresent(
+        (key) -> writeOptions.add(BlobWriteOption.encryptionKey(key)));
+    gcpProperties.userProject().ifPresent(
+        (userProject) -> writeOptions.add(BlobWriteOption.userProject(userProject)));
+
+    WriteChannel channel = storage.writer(BlobInfo.newBuilder(blobId).build(),
+        writeOptions.toArray(new BlobWriteOption[0]));
+
+    gcpProperties.channelWriteChunkSize().ifPresent(channel::setChunkSize);
```

Comment:
can we add some basic checks for configuration such as check size > 0 if present.

Reply:
I feel like we shouldn't be validating on top of the channel.  Also, knowing a little bit about how the implementation works, this is more of a suggestion than an explicit set.  The channel has a min size it will use regardless of what you set.

Reply:
I guess I should clarify that it valid to use a negative number to effectively say "use the min buffer size allowed".  While that may be bad practice, I think adding additional checks here is unnecessary.

Reply:
I see, thanks for the explanation!